flask/common/src/main/java/net/corda/flask/common/Flask.java

```diff
@@ -0,0 +1,124 @@
+package net.corda.flask.common;
+
+import lombok.SneakyThrows;
+
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.math.BigInteger;
+import java.security.DigestInputStream;
+import java.security.MessageDigest;
+import java.util.AbstractMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Supplier;
+import java.util.zip.CRC32;
+import java.util.zip.ZipEntry;
+
+public class Flask {
+
+    public static class Constants {
+        public static final String DEFAULT_LAUNCHER_NAME = "net.corda.flask.launcher.Launcher";
+        public static final String LIBRARIES_FOLDER = "LIB-INF";
+        public static final String METADATA_FOLDER = "META-INF";
+        public static final int BUFFER_SIZE = 0x10000;
+    }
+
+    public static class ManifestAttributes {
+        public static final String APPLICATION_CLASS = "Application-Class";
+        public static final String JVM_ARGS = "JVM-Args";
+        public static final String JAVA_AGENTS = "Java-Agents";
```

Comment:
We don't want to parse "lists of things from strings". This complexity should be entirely avoidable.

Reply:
sorry, I don't understand what you mean...

Reply:
These are `String` attributes:
```java
String jvmArgsManifestString = manifest.getMainAttributes().getValue(Flask.ManifestAttributes.JVM_ARGS);
```
and
```java
String javaAgentsManifestString = manifest.getMainAttributes().getValue(Flask.ManifestAttributes.JAVA_AGENTS);
```
which presumably contain multiple values. We can avoid needing to parse these values into `List<String>` (which is what `ProcessBuilder` requires) by storing them as individual values in the first place.

Reply:
there is no way to store a list of arbitrary string values in the manifest without escaping it and even if there was it wouldn't give any advantage since the escaping code is still needed when providing JVM arguments from command line with 
```
java -Dnet.corda.flask.jvm.args="-Dsome.property=\"some value\nwith newlines\"" -jar flask.jar
```

Reply:
Why does this information need to go into the manifest in the first place? You "own" the flask's `META-INF` directory and so can store whatever you like in there using a more convenient format.

As for `-Dnet.corda.flask.jvm.args=`, and assuming we still need this feature (seeing as we can store this information in `node.conf` instead), I don't see why we need to use a system property. For example:
```java
public class CommandArgs {
    public static void main(String[] args) {
        for (int i = 0; i < args.length; ++i) {
            System.out.println("ARG[" + i + "]=[" + args[i] + ']');
        }
    }
}

$ java CommandArgs -flaskJvmArg=-Dfoo=bar -flaskJvmArg='-XX:+UseG1GC' -flaskJvmArg=-Dcorda.thingy="one two three" dogfood
ARG[0]=[-flaskJvmArg=-Dfoo=bar]
ARG[1]=[-flaskJvmArg=-XX:+UseG1GC]
ARG[2]=[-flaskJvmArg=-Dcorda.thingy=one two three]
ARG[3]=[dogfood]
```
So any `-flaskJvmArg=<burble>` argument would be removed, with the `<burble>` value taken _verbatim_ as a new JVM argument. And everything else would be passed straight through to the application of course.

Parsing command lines is tricky, so why not let the command shell do the hard work for you?

I used Linux/bash for this trivial example; it would be worth seeing what it does on Windows too.

Reply:
ok, this sound feasible.. I'll do this to pass JVM argument from the cli and store the others in a property file under `META-INF` in the form
```
0=-Dfoo=bar
1=-XX:+UseG1GC
2=-Dcorda.thingy\=one two three
```

Reply:
OK, thanks. @JamesHR3, if you run Windows (I don't think @woggioni does any more than I do...) can you compile the `CommandArgs` class above and tell us what the output from that command line is please?

Reply:
Output looks the same to me:
```shell
java CommandArgs -flaskJvmArg=-Dfoo=bar -flaskJvmArg='-XX:+UseG1GC' -flaskJvmArg=-Dcorda.thingy="one two three" dogfood
ARG[0]=[-flaskJvmArg=-Dfoo=bar]
ARG[1]=[-flaskJvmArg='-XX:+UseG1GC']
ARG[2]=[-flaskJvmArg=-Dcorda.thingy=one two three]
ARG[3]=[dogfood]
```

Reply:
Not _quite_ - `ARG[1]` still has its single quotes on Windows :detective:.

Reply:
yeah, but that is because the Windows shell is not a posix shell and it will probably differ between Windows Power Shell and cmd.exe.. there is most likely a way to put whitespaces in Windows without splitting command line arguments

Reply:
it worked for `-flaskJvmArg=-Dcorda.thingy="one two three"` so we can just tell people they should use double quotes

Reply:
it would be intersting to test it with something more challenging, like a string that contains double and single quotes (meant to forwared to the JVM) and (more challenging) a newline (in a posix shell that is `$'\n'` outside of both singe and double quotes)

Reply:
The idea is _not_ to be too complicated :wink:. All we need is for there to be _*a*_ way to pass these values via the command line that works for both UNIX and Windows, Wrapping `-flaskJvmArg=<burble>` _completely_ in quotes would be enough. We are still expecting the _really_ complicated stuff to be added to `node.conf`.

Reply:
> I'm currently thinking "Remove -flaskJvmArg= and then check if the remainder is between a matching pair of quotes.".
what do you mean? double quotes won't be forwarded to the JVM (unless they are escaped)

Reply:
From `gradlew.bat`:
```
@rem Execute Gradle
"%JAVA_EXE%" %DEFAULT_JVM_OPTS% %JAVA_OPTS% %GRADLE_OPTS% "-Dorg.gradle.appname=%APP_BASE_NAME%" -classpath "%CLASSPATH%" org.gradle.wrapper.GradleWrapperMain %*
```
so wrapping each argument in quotes is "A Thing" on Windows.

Reply:
generally speaking, this problem of shell escaping has nothing to do with this code, I expect the user to be able to somehow use an arbitrary string as a process command line argument, my role is simply to be able to correctly split that string in a consistent and reliable way into multiple arguments for the new Java subprocess, this can be achieved escaping the whitespaces in the string (as I did before) or asking the user to split those arguments himself prepending them with a known magic prefix as you suggested, it isn't then relevant if on Linux single quotes are interpreted by the shell and hence removed from the string while on Windows are not.

Reply:
The point about using an argument prefix like `-flaskJvmArg=` is that it allows the command shell to do the hard work of parsing the individual arguments for you. Even your Gradle task has
```java
ListProperty<String> jvmArgs;
```
and so these are already delimited for you too. So given that you start with `List<String>`, and need to pass `List<String>` to `ProcessBuilder`, does it really make sense to concatenate them all together into a single `String` as an intermediate step?